Twitter API dependencies (do not need to run)

In [ ]:
# Twitter Handle Source: https://pressgallery.house.gov/member-data/members-official-twitter-handles

#Import appropriate libraries

import tweepy
import pandas as pd

#Define Access Keys

api_key = 't7U9rmjmbvqZCQnherHkrDZ9q'
api_key_secret = 'k5FwX8KGxruNp9oUlhNCb1xHtJ1JjIIlI7ozai5vGaMM2pIaCU'

access_token = '1532526934269239303-nQpMFHharyoQCfc8FoNuUwAUFdJM6Y'
access_token_secret = 'ndON5Ktl0MEOFDtjGMjGVkuRz4D6FWpaPruh2IHUWEgC0'

# authentication
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

api = tweepy.API(auth)

# Democratic Rep Tweets (do not need to run)

In [ ]:
#Pull tweet based off user

users = ["RepDonBeyer","SanfordBishop","repblumenauer","RepLBR","RepBonamici","RepBourdeaux","RepBowman","CongBoyle","RepAnthonyBrown","RepShontelBrown","RepBrownley","RepCori","RepCheri","GKButterfield","RepCarbajal","RepCardenas","RepAndreCarson","RepTroyCarter","RepCartwright","RepEdCase","RepCasten","USRepKCastor","JoaquinCastrotx","CongresswomanSC","RepJudyChu","davidcicilline","RepCicilline","RepKClark","RepYvetteClarke","repcleaver","WhipClyburn","RepCohen","GerryConnolly","repjimcooper","RepLouCorrea","RepJimCosta","RepJoeCourtney","RepAngieCraig","RepCharlieCrist","RepJasonCrow","RepCuellar","RepDavids","RepDannyDavis","RepDean","RepPeterDeFazio","RepDianaDeGette","rosadelauro","RepDelBene","repdelgado","RepValDemings","RepDeSaulnier","RepTedDeutch","RepDebDingell","RepLloydDoggett","USRepMikeDoyle","RepEscobar","RepAnnaEshoo","RepEspaillat","RepDwightEvans","RepFletcher","RepBillFoster","RepLoisFrankel","RepRubenGallego","RepGaramendi","RepSylviaGarcia"]
limit= 500
data = []
for user in users:
  tweets = tweepy.Cursor(api.user_timeline, screen_name=user, count=200, tweet_mode='extended').items(limit)

# tweets = api.user_timeline(screen_name=user, count=limit, tweet_mode='extended')

# create DataFrame
  columns = ['User', 'Tweet']
  for tweet in tweets:
        print(tweet.user.screen_name)
        data.append([tweet.user.screen_name, tweet.full_text])
        df_1 = pd.DataFrame(data, columns=columns)

#Export to csv
from google.colab import files
df_1.to_csv('filename.csv') 
files.download('filename.csv')

RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepDonBeyer
RepD

KeyboardInterrupt: ignored

# Republican Tweets (do not need to run)

In [ ]:
#Pull tweet based off user


users = ["RepBice","RepAndyBiggsAZ","RepGusBilirakis","RepDanBishop","RepBoebert","RepBost","RepKevinBrady","RepMoBrooks","VernBuchanan","RepKenBuck","RepLarryBucshon","RepTedBudd","RepTimBurchett","michaelcburgess","KenCalvert","RepKatCammack","RepMikeCarey","RepJerryCarl","RepBuddyCarter","JudgeCarter","RepCawthorn","RepSteveChabot","RepLizCheney","RepBenCline","RepCloudTX","Rep_Clyde","TomColeOK04","RepJamesComer","RepRickCrawford","RepDanCrenshaw","RepJohnCurtis","WarrenDavidson","RodneyDavis","DesJarlaisTN04","MarioDB","RepDonaldsPress","RepJeffDuncan","DrNealDunnFL2","RepEllzey","RepTomEmmer","RepRonEstes","RepPatFallon","RepFeenstra","RepDrewFerguson","RepFischbach","RepFitzgerald","RepBrianFitz","RepChuck","JeffFortenberry","virginiafoxx","RepFranklin","RepRussFulcher","RepMattGaetz","RepGallagher","RepGarbarino","RepMikeGarcia","RepBobGibbs","RepCarlos",]
limit= 500
data = []
for user in users:
  tweets = tweepy.Cursor(api.user_timeline, screen_name=user, count=200, tweet_mode='extended').items(limit)

# tweets = api.user_timeline(screen_name=user, count=limit, tweet_mode='extended')

# create DataFrame
  columns = ['User', 'Tweet']
  for tweet in tweets:
        print(tweet.user.screen_name)
        data.append([tweet.user.screen_name, tweet.full_text])
        df_1 = pd.DataFrame(data, columns=columns)

  print(df_1)

from google.colab import files
df_1.to_csv('filename.csv') 
files.download('filename.csv')


Streaming output truncated to the last 5000 lines.
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx
virginiafoxx


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# RUN THIS! Final Dataframe - Copy folder "Capstone_ArkeNews_Corpus" to your Google drive

dataframe name = df_final

In [ ]:
#Mount Personal Google Drive

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Import corpus csv

import pandas as pd

df_final = pd.read_csv("/content/drive/My Drive/Capstone_ArkeNews_Corpus/Corpus_USSenatorTweets.csv")

print(df_final.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72218 entries, 0 to 72217
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   User    72217 non-null  object
 1   Tweet   72218 non-null  object
 2   Label   72218 non-null  object
dtypes: object(3)
memory usage: 1.7+ MB
None


In [ ]:
df_final.head()

User                                              Tweet Label
0  RepAdams  I didn't ask for a pardon.\n\nBut a certain Pr...     D
1  RepAdams  Your daily reminder that instead of saying som...     D
2  RepAdams  I’m proud to work with my Congressional collea...     D
3  RepAdams  RT @grantstern: Multiple Republican Congressme...     D
4  RepAdams  RT @MacFarlaneNews: Approx. 140 police officer...     D

# Next Steps for team: 



1.   Clean corpus - Done
2.   Apply Step 2 and 3 from ArkeNews Model



In [ ]:

#define a function to clean up the tweets. input - text field of all #the rows, output - cleaned text - remove @, hashtags, urls, emojis 

import re 

def cleanUpTweet(txt):
    # Remove mentions
    txt = re.sub(r'@+', '', txt)
    # Remove hashtags
    txt = re.sub(r'#[A-Z0-9]+', '', txt)
    # Remove retweets:
    txt = re.sub(r'RT : ', '', txt)
    # Remove urls
    txt = re.sub(r'https?:\/\/[A-Za-z0-9\.\/]+', '', txt)
    #remove amp
    txt = re.sub(r'&amp;', '', txt)
    #rempve strange characters
    txt = re.sub(r'ðŸ™', '', txt)
    #remove new lines
    txt = re.sub(r'\n', ' ', txt)
    return txt

df_final['Cleaned_Tweet'] = df_final['Tweet'].apply(cleanUpTweet)

In [ ]:
df_final.head()

User                                              Tweet Label  \
0  RepAdams  I didn't ask for a pardon.\n\nBut a certain Pr...     D   
1  RepAdams  Your daily reminder that instead of saying som...     D   
2  RepAdams  I’m proud to work with my Congressional collea...     D   
3  RepAdams  RT @grantstern: Multiple Republican Congressme...     D   
4  RepAdams  RT @MacFarlaneNews: Approx. 140 police officer...     D   

                                       Cleaned_Tweet  
0  I didn't ask for a pardon.  But a certain Pres...  
1  Your daily reminder that instead of saying som...  
2  I’m proud to work with my Congressional collea...  
3  RT grantstern: Multiple Republican Congressmen...  
4  RT MacFarlaneNews: Approx. 140 police officers...

# Classification Model 1 - Democratic/Other

## Encoding labels

In [ ]:
#Encode Label: D = 1 / R = 0

df_final['Label_Number'] = df_final.Label.map({'R':0, 'D':1})
df_final.head()

User                                              Tweet Label  \
0  RepAdams  I didn't ask for a pardon.\n\nBut a certain Pr...     D   
1  RepAdams  Your daily reminder that instead of saying som...     D   
2  RepAdams  I’m proud to work with my Congressional collea...     D   
3  RepAdams  RT @grantstern: Multiple Republican Congressme...     D   
4  RepAdams  RT @MacFarlaneNews: Approx. 140 police officer...     D   

                                       Cleaned_Tweet  Label_Number  
0  I didn't ask for a pardon.  But a certain Pres...             1  
1  Your daily reminder that instead of saying som...             1  
2  I’m proud to work with my Congressional collea...             1  
3  RT grantstern: Multiple Republican Congressmen...             1  
4  RT MacFarlaneNews: Approx. 140 police officers...             1

In [ ]:
#Tweet Count

df_final.groupby('Label_Number').describe()

User                               Tweet         \
              count unique              top freq  count unique   
Label_Number                                                     
0             35522     72  Robert_Aderholt  500  35523  34891   
1             36695     75         RepAdams  500  36695  35930   

                                                                      Label  \
                                                            top freq  count   
Label_Number                                                                  
0             RT @RepFranklin: The creation of a "Disinforma...    8  35523   
1             RT @rosadelauro: The infant formula shortage c...   16  36695   

                               Cleaned_Tweet                  
             unique top   freq         count unique top freq  
Label_Number                                                  
0                 1   R  35523         35523  34489      185  
1                 1   D  36695         36695  35679      121

36,695 tweets are Democratic
35,523 tweets are not democratic.

50% split.

In [ ]:
##Additional EDA - Sentence Length

## **Model**

### Train / Validation Split

In [ ]:
# Train Val Split - 75% Train, 25% Val
from sklearn.model_selection import train_test_split

X = df_final['Cleaned_Tweet']
y = df_final['Label_Number']


X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=42)

NameError: ignored

In [ ]:
print(X_train.shape)
print(X_val.shape)
print(y_train.shape)
print(y_val.shape)

(54163,)
(18055,)
(54163,)
(18055,)


### Part 1.a: Preprocessing and FE Pipeline

In [ ]:
#Count Vectorization 

from sklearn.feature_extraction.text import CountVectorizer

vect = CountVectorizer()
vect.fit(X_train)

CountVectorizer()

In [ ]:
# learn training data vocabulary, then use it to create a document-term matrix
X_train_dtm = vect.transform(X_train)

In [ ]:
# combine fit and transform into a single step
X_train_dtm = vect.fit_transform(X_train)

In [ ]:
# examine the document-term matrix
X_train_dtm

<54163x45841 sparse matrix of type '<class 'numpy.int64'>'
	with 1365588 stored elements in Compressed Sparse Row format>

In [ ]:
# transform testing data (using fitted vocabulary) into a document-term matrix
X_val_dtm = vect.transform(X_val)
X_val_dtm

<18055x45841 sparse matrix of type '<class 'numpy.int64'>'
	with 449561 stored elements in Compressed Sparse Row format>

In [ ]:
#TFID 

from sklearn.feature_extraction.text import TfidfTransformer

tfidf_transformer = TfidfTransformer()
tfidf_transformer.fit(X_train_dtm)
tfidf_transformer.transform(X_train_dtm)

<54163x45841 sparse matrix of type '<class 'numpy.float64'>'
	with 1365588 stored elements in Compressed Sparse Row format>

### Part 1.b: Model Training/Tuning/Cross Validation 

In [ ]:
#Multinomial Naive Bayes

from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()

In [ ]:
%time nb.fit(X_train_dtm, y_train)

CPU times: user 27.8 ms, sys: 780 µs, total: 28.6 ms
Wall time: 29.3 ms


MultinomialNB()

In [ ]:
#Prediction

y_pred_class = nb.predict(X_val_dtm)

from sklearn import metrics
metrics.accuracy_score(y_val, y_pred_class)

0.856272500692329

In [ ]:
metrics.confusion_matrix(y_val, y_pred_class)

array([[7350, 1490],
       [1105, 8110]])

In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.pipeline import Pipeline

pipe = Pipeline([('bow', CountVectorizer()), 
                 ('tfid', TfidfTransformer()),  
                 ('model', MultinomialNB())])
pipe.fit(X_train, y_train)

Pipeline(steps=[('bow', CountVectorizer()), ('tfid', TfidfTransformer()),
                ('model', MultinomialNB())])

In [ ]:
y_pred = pipe.predict(X_val)

In [ ]:
metrics.accuracy_score(y_val, y_pred)

0.8549432290224315

# Step 2: Common Phrases

Stop word removal

*   Stop Word Removal
*   Text Vectorization


In [ ]:
import nltk
nltk.download('stopwords')
from sklearn.feature_extraction.text import CountVectorizer
from nltk.corpus import stopwords
nltk_stop_words = stopwords.words('english')

tf_vectorizer = CountVectorizer(max_df=0.8, min_df=50,
    ngram_range = (1,2),
    binary=True,
    stop_words=nltk_stop_words)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
term_frequencies = tf_vectorizer.fit_transform(df_final)

ValueError: ignored

In [ ]:
phrases_df = pd.DataFrame(data=tf_vectorizer.get_feature_names(),columns=['phrase'])
phrases_df['total_occurrences']=term_frequencies.sum(axis=0).T
phrases_df.sort_values(by='total_occurrences',ascending=False).head(20).to_csv('top_20_overall.csv',index=False)

# Step 3: Politically Charged Phrases



In [ ]:
dem_tfs = tf_vectorizer.transform(texts_df[texts_df.party=='Democrat'].text.tolist())
rep_tfs = tf_vectorizer.transform(texts_df[texts_df.party=='Republican'].text.tolist())

In [ ]:
bias = (p_rep - p_dem) / (p_rep + p_dem)